Bug Fix for Issue #349, menus longer than the window height

timurvural opened the pull request:
Bug Fix for issue #349.

Originally, keepWithin was called on the menu when it extended past the boundaries of its world, which simply shifted it up in the world to keep it visible. When the menu is taller than the window, however, some parts of the menu become inaccessible, as shown below.

![image](https://cloud.githubusercontent.com/assets/5465011/11166513/dfa571fc-8aea-11e5-910e-6c3183b828f9.png)

I wrote a new method keepVisible, which encloses the menu in a scrollFrameMorph when it extends past the world boundary. This makes the rest of the menu accessible by scrolling. I also changed Handmorph's processMouseDown method to make sure the menu stays selected when its scrollFrameMorph is clicked on. An example of a large menu made scrollable by its scrollFrameMorph is below.

<img width="1438" alt="screen shot 2015-11-14 at 4 16 23 pm" src="https://cloud.githubusercontent.com/assets/5465011/11166520/6f2ca336-8aeb-11e5-9a2a-7aa31b400c24.png">

jmoenig commented:
very cool! And very timely. Look, what I've been working on myself this past week in GP: Scrollable Menus! :-)

<img width="230" alt="scrollable_menu_gp" src="https://cloud.githubusercontent.com/assets/2596285/11168125/a67bf750-8b82-11e5-8c34-77e3f0efde05.png">

Looks just like yours :-)

So, it appears like you're not changing the internal structure of Menus, but instead just always wrap a scroll frame around the whole menu, regardless of whether it all fits onscreen. Did I get this right from just glancing at your code? How do you handle menu labels? Do you start scrolling underneath them (so the label stays always visible), like this:

<img width="588" alt="scrollable_menu_gp_2" src="https://cloud.githubusercontent.com/assets/2596285/11168138/7cd9bd78-8b83-11e5-8432-ffdc973bf3c0.png">

or do you scroll everything?

Also, looking at your screenshot it seems that in your code menus are loosing their slight drop-shadows and their surrounding frames (because the surrounding scroll frame clips them). Is that right?

bb010g commented:
What's GP?

timurvural commented:
That's right - rather than making the menus themselves scrollable, I wrapped the menus inside a scroll frame. The menu only gets wrapped inside a scroll frame if it can't fit on the screen without being repositioned. 

Menu labels can be scrolled, so they are not always visible. 

Here is a different screenshot for the shadows - they are still visible, but only when they are still within the scroll frame.

<img width="1440" alt="screen shot 2015-11-16 at 12 01 31 am" src="https://cloud.githubusercontent.com/assets/5465011/11176823/7153f820-8bf5-11e5-8c91-f6897b0abafc.png">

In the screenshots you mentioned, the menu was not originally inside a frame, so there was no frame to clip. If the menu had started off already contained in a frame, however, that frame would have gotten clipped.

khotchkiss13 commented:
Like @bb010g, I too am curious as to what GP is

jmoenig commented:
GP is a new general purpose blocks programming language currently under development by John Maloney, Yoshiki Ohshima and myself. It's still in early alpha, we did show it off at this year's Scratch conference in Amsterdam and at a recent VL/HCC "Blocks & Beyond" workshop in Atlanta. Here are two things to read about it: http://dspace.mit.edu/handle/1721.1/98451 and http://www.vpri.org/pdf/tr2015003_modsys.pdf and there's also an article in the German Scratch wiki: http://scratch-dach.info/wiki/GP_(Programmiersprache)

timurvural commented:
I've made some updates. The scroll frame is now created inside the menu rather than outside, and the menu's contents are added to the scroll frame. The scroll frame now won't clip the menu's surrounding frame or the menu's shadow. 

<img width="1439" alt="screen shot 2015-12-04 at 4 43 33 pm" src="https://cloud.githubusercontent.com/assets/5465011/11604861/21f00a42-9aa7-11e5-961c-c030710e5eb8.png">


Also, if the menu is short enough to fit on screen, the menu is simply shifted up so it stays visible.

<img width="1438" alt="screen shot 2015-12-04 at 4 51 26 pm" src="https://cloud.githubusercontent.com/assets/5465011/11604867/4cf7f6be-9aa7-11e5-907f-e368a2a1e65e.png">


If the menu has a label, the label stays pinned to the top of the menu.

<img width="1440" alt="screen shot 2015-12-04 at 4 52 46 pm" src="https://cloud.githubusercontent.com/assets/5465011/11604877/7bd47caa-9aa7-11e5-80f5-f2201d12a535.png">

jmoenig commented:
Nice! Thank you!

liam4 commented:
Awesome! I was having this issue earlier when trying to look through a very long list.

khotchkiss13 commented:
This looks awesome!

cycomachead commented on morphic.js:
```diff
@@ -6694,6 +6694,7 @@ MenuMorph.prototype.init = function (target, title, environment, fontSize) {
     this.isListContents = false;
     this.hasFocus = false;
     this.selection = null;
+    this.scrollMenu=null;
```
Spacing, use `x = y` for `=` stuff

cycomachead commented on morphic.js:
```diff
@@ -6903,18 +6904,37 @@ MenuMorph.prototype.unselectAllItems = function () {
-    world.activeMenu = this;
     this.world = world; // optionally enable keyboard support
+    world.activeMenu = this;
+    if (this.scrollMenu != undefined && this.scrollMenu != null) {
```
I think this should probably just be `if (this.scrollMenu) {`

cycomachead commented on morphic.js:
```diff
@@ -6903,18 +6904,37 @@ MenuMorph.prototype.unselectAllItems = function () {
+    if (this.scrollMenu != undefined && this.scrollMenu != null) {
+        pos = this.position();
+        var labelHeight = 0;
+        if (this.label != null) {
```
For all comparisons you should use `===` and `!==`

cycomachead commented on morphic.js:
```diff
@@ -6903,18 +6904,37 @@ MenuMorph.prototype.unselectAllItems = function () {
+        this.bounds.corner.y = Math.round(0.98 * world.height());
+        this.scrollMenu.setWidth(this.width() - 6);
+        this.scrollMenu.setHeight(this.bottom() - this.scrollMenu.top() - 4);
+        for (var i = this.children.length ; i > -1 ; i += -1){
```
Unless there's a specific need to iterate backwards, this should be `this.children.forEach()` style loop.

timurvural replied:
In this case, I iterated backwards because members of this.children are getting removed during iteration.

cycomachead commented on morphic.js:
```diff
@@ -6944,6 +6964,16 @@ MenuMorph.prototype.popUpCenteredInWorld = function (world) {
     );
 };
 
+MenuMorph.prototype.keepVisible=function(world){
```
Update the spacing here.

cycomachead commented on morphic.js:
```diff
@@ -7040,6 +7070,10 @@ MenuMorph.prototype.destroy = function () {
     if (this.hasFocus) {
         this.world.keyboardReceiver = null;
     }
+    if (this.scrollMenu != null) {
```
`!this.scrollMenu`

cycomachead commented on morphic.js:
```diff
@@ -9566,9 +9600,10 @@ HandMorph.prototype.processMouseDown = function (event) {
+                ) && !contains(morph.allParents(), this.world.activeMenu.scrollMenu)) {
+
+                this.world.activeMenu.destroy();}
+            else {
```
Spacing on these lines

cycomachead commented:
I added a bunch of comments for code style, this looks great!

The only bug I see is that the top and bottom of the scroll bars are cut off, but I don't immediately see why.....

cycomachead commented:
Also, it looks like there should be a teeny bit of padding between the right edge of the menu and a scrollbar.

cycomachead commented on morphic.js:
```diff
@@ -6903,18 +6904,37 @@ MenuMorph.prototype.unselectAllItems = function () {
+    world.activeMenu = this;
+    if (this.scrollMenu != undefined && this.scrollMenu != null) {
+        pos = this.position();
+        var labelHeight = 0;
```
This should be the first line of the function.

timurvural commented:
Thanks for catching all of that! I've updated the style of the code and added a little bit of padding. It turned out the scroll button wasn't being clipped - just had to make the slider more transparent so the button could show better.

timurvural commented:
@jmoenig, Would you mind taking another look at this? I've fixed all the issues that were pointed out in this thread.

jmoenig commented:
@timurvural sure, thanks!

brianharvey commented:
I'm confused, was this pulled?

cycomachead commented:
No, but the issue was resolved by Jens building the same thing. :)
